You're right, thanks. I should have noticed that the argument causing the warning was [a call to `net_ntohl()`](https://github.com/zephyrproject-rtos/zephyr/blob/74a32bdff3ca6f8ba1acaae05b223488fbbc81af/subsys/net/lib/dhcpv4/dhcpv4.c#L1605-L1609), which resolves to `BSWAP_32()` on little-endian architectures like Xtensa.

The CI warning in question, from `xtensa-espressif_esp32_zephyr-elf-gcc` in Zephyr SDK 1.0.1:
```
In file included from /__w/zephyr/zephyr/include/zephyr/logging/log.h:11,
                 from /__w/zephyr/zephyr/subsys/net/lib/dhcpv4/dhcpv4.c:13:
/__w/zephyr/zephyr/subsys/net/lib/dhcpv4/dhcpv4.c: In function 'net_dhcpv4_input':
/__w/zephyr/zephyr/include/zephyr/net/net_log.h:27:35: error: format '%x' expects argument of type 'unsigned int', but argument 6 has type 'long unsigned int' [-Werror=format=]
   27 | #define NET_DBG(fmt, ...) LOG_DBG("(%s): " fmt,                         \
      |                                   ^~~~~~~~
/__w/zephyr/zephyr/include/zephyr/logging/log_core.h:336:50: note: in definition of macro 'Z_LOG2'
  336 |                         z_log_printf_arg_checker(__VA_ARGS__);                                     \
      |                                                  ^~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/logging/log.h:93:22: note: in expansion of macro 'Z_LOG'
   93 | #define LOG_DBG(...) Z_LOG(LOG_LEVEL_DBG, __VA_ARGS__)
      |                      ^~~~~
/__w/zephyr/zephyr/include/zephyr/net/net_log.h:27:27: note: in expansion of macro 'LOG_DBG'
   27 | #define NET_DBG(fmt, ...) LOG_DBG("(%s): " fmt,                         \
      |                           ^~~~~~~
/__w/zephyr/zephyr/subsys/net/lib/dhcpv4/dhcpv4.c:1605:9: note: in expansion of macro 'NET_DBG'
 1605 |         NET_DBG("Received dhcp msg [op=0x%x htype=0x%x hlen=%u xid=0x%x "
      |         ^~~~~~~
/__w/zephyr/zephyr/include/zephyr/net/net_log.h:27:35: error: format '%x' expects argument of type 'unsigned int', but argument 5 has type 'long unsigned int' [-Werror=format=]
   27 | #define NET_DBG(fmt, ...) LOG_DBG("(%s): " fmt,                         \
      |                                   ^~~~~~~~
/__w/zephyr/zephyr/include/zephyr/logging/log_core.h:336:50: note: in definition of macro 'Z_LOG2'
  336 |                         z_log_printf_arg_checker(__VA_ARGS__);                                     \
      |                                                  ^~~~~~~~~~~
/__w/zephyr/zephyr/include/zephyr/logging/log.h:93:22: note: in expansion of macro 'Z_LOG'
   93 | #define LOG_DBG(...) Z_LOG(LOG_LEVEL_DBG, __VA_ARGS__)
      |                      ^~~~~
/__w/zephyr/zephyr/include/zephyr/net/net_log.h:27:27: note: in expansion of macro 'LOG_DBG'
   27 | #define NET_DBG(fmt, ...) LOG_DBG("(%s): " fmt,                         \
      |                           ^~~~~~~
/__w/zephyr/zephyr/subsys/net/lib/dhcpv4/dhcpv4.c:1626:17: note: in expansion of macro 'NET_DBG'
 1626 |                 NET_DBG("Unexpected op (%d), xid (%x vs %x) or chaddr",
      |                 ^~~~~~~
```

The warning is quite odd, since GCC [documents](https://gcc.gnu.org/onlinedocs/gcc/Byte-Swapping-Builtins.html) `__builtin_bswap32()` to return `uint32_t`, and several other `uint32_t`s formatted with `%x` specifiers in the same call don't produce warnings.

I think I found the cause, though: [as Espressif notes](https://docs.espressif.com/projects/esp-idf/en/v5.0/esp32/migration-guides/release-5.x/gcc.html#int32-t-and-uint32-t-for-xtensa-compiler), upstream GCC uses `unsigned long` for `uint32_t` on Xtensa, so that's what `__builtin_bswap32()` returns. However, the Zephyr SDK takes `stdint.h` from picolibc, [not from GCC](https://github.com/zephyrproject-rtos/gcc/commit/4be5217034ef47d3c5dc84caf3a2f40b51b4785a). picolibc has [different logic](https://github.com/zephyrproject-rtos/picolibc/blob/01254932e8e81085817ed61fd858648584ffe37c/newlib/libc/include/machine/_default_types.h#L169-L182) to determine the type of `uint32_t` and uses `unsigned int` on Xtensa. This typically doesn't cause problems since they're the same size, but it means the `uint32_t` returned by builtins isn't the same one named by typing `uint32_t`.

I'm not 100% sure of this explanation, since GCC's internals are hard to follow, but it's consistent with the symptom we see. Your proposed change works around the issue, and I'll update the PR with it momentarily.

--- a/include/zephyr/sys/byteorder.h
+++ b/include/zephyr/sys/byteorder.h
@@ -18,14 +18,22 @@
 #include <zephyr/sys/util_macro.h>
 #include <zephyr/toolchain.h>
 
+#if HAS_BUILTIN(__builtin_bswap16)
+#define BSWAP_16(x) ((uint16_t) __builtin_bswap16(x))
+#else
 #define BSWAP_16(x) ((uint16_t) ((((x) >> 8) & 0xff) | (((x) & 0xff) << 8)))
+#endif
 #define BSWAP_24(x) ((uint32_t) ((((x) >> 16) & 0xff) | \
 				   (((x)) & 0xff00) | \
 				   (((x) & 0xff) << 16)))
+#if HAS_BUILTIN(__builtin_bswap32)
+#define BSWAP_32(x) ((uint32_t) __builtin_bswap32(x))
+#else
 #define BSWAP_32(x) ((uint32_t) ((((x) >> 24) & 0xff) | \
 				   (((x) >> 8) & 0xff00) | \
 				   (((x) & 0xff00) << 8) | \
 				   (((x) & 0xff) << 24)))
+#endif
 #define BSWAP_40(x) ((uint64_t) ((((x) >> 32) & 0xff) | \
 				   (((x) >> 16) & 0xff00) | \
 				   (((x)) & 0xff0000) | \
@@ -37,6 +45,9 @@
 				   (((x) & 0xff0000) << 8) | \
 				   (((x) & 0xff00) << 24) | \
 				   (((x) & 0xff) << 40)))
+#if HAS_BUILTIN(__builtin_bswap64)
+#define BSWAP_64(x) ((uint64_t) __builtin_bswap64(x))
+#else
 #define BSWAP_64(x) ((uint64_t) ((((x) >> 56) & 0xff) | \
 				   (((x) >> 40) & 0xff00) | \
 				   (((x) >> 24) & 0xff0000) | \
@@ -45,6 +56,7 @@
 				   (((x) & 0xff0000) << 24) | \
 				   (((x) & 0xff00) << 40) | \
 				   (((x) & 0xff) << 56)))
+#endif
 
 /** @def sys_le16_to_cpu
  *  @brief Convert 16-bit integer from little-endian to host endianness.

--- a/include/zephyr/toolchain/gcc.h
+++ b/include/zephyr/toolchain/gcc.h
@@ -332,6 +332,13 @@ do {                                                                    \
 #define HAS_BUILTIN___builtin_mul_overflow 1
 #define HAS_BUILTIN___builtin_div_overflow 1
 #endif
+#if TOOLCHAIN_GCC_VERSION >= 40800
+#define HAS_BUILTIN___builtin_bswap16 1
+#endif
+#if TOOLCHAIN_GCC_VERSION >= 40300
+#define HAS_BUILTIN___builtin_bswap32 1
+#define HAS_BUILTIN___builtin_bswap64 1
+#endif
 #if __GNUC__ >= 4
 #define HAS_BUILTIN___builtin_clz 1
 #define HAS_BUILTIN___builtin_clzl 1

